File: 프로그래머스/Level 1/숫자 짝꿍.py
# https://school.programmers.co.kr/learn/courses/30/lessons/131128
def solution(X, Y):
    answer = ""
    tmpArr = []
    xDict = {}
    yDict = {}

    for x in X:
        if not x in xDict:
            xDict[x] = 1
        else:
            xDict[x] += 1

    for y in Y:
        if not y in yDict:
            yDict[y] = 1
        else:
            yDict[y] += 1

    # print(xDict)
    # print(yDict)

    for (key, value) in xDict.items():
        if key in yDict:
            minLen = min(xDict[key], yDict[key])
            for _ in range(minLen):
                tmpArr.append(key)
            # print(minLen)
    # print(tmpArr)

    if not tmpArr:
        return "-1"
    tmpArr.sort(reverse=True)
    answer = "".join(tmpArr)
    return answer if answer[0] != "0" else "0"

File: 프로그래머스/Level 1/숫자 짝꿍.cpp
// https://school.programmers.co.kr/learn/courses/30/lessons/131128
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

bool compare(char a, char b) {
    return a > b;
}

string solution(string X, string Y) {
    string answer = "";
    unordered_map<char,int> xMap;
    unordered_map<char,int> yMap;
    vector<char> tmp;
    
    for (char& x: X)
    {
        xMap[x] += 1; 
    }
    
    for (char& y: Y)
    {
        yMap[y] += 1; 
    }
    
    for (auto iter = xMap.begin(); iter != xMap.end(); ++iter)
    {
        if (yMap.find(iter->first) != yMap.end())
        {
            int minCount = min(iter->second, yMap[iter->first]);
            while (minCount--) answer += iter->first;
        }
    }
    
    if (!answer.size()) answer = "-1";
    else
    {
        sort(answer.begin(), answer.end(), compare);
        if (answer[0] == '0') answer = "0";
    }
    
    return answer;
}